stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/channel/internal/DatabaseChannelRepository.kt

```diff
@@ -16,52 +16,64 @@
+import android.util.LruCache
 import io.getstream.chat.android.client.persistance.repository.ChannelRepository
+import io.getstream.chat.android.core.utils.date.maxOf
+import io.getstream.chat.android.core.utils.date.minOf
 import io.getstream.chat.android.models.Channel
 import io.getstream.chat.android.models.Member
 import io.getstream.chat.android.models.Message
 import io.getstream.chat.android.models.User
-import io.getstream.chat.android.offline.repository.domain.channel.lastMessageInfo
-import io.getstream.chat.android.offline.repository.domain.channel.member.internal.toEntity
-import io.getstream.chat.android.offline.repository.domain.channel.member.internal.toModel
 import io.getstream.log.taggedLogger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
 import java.util.Date
 
 /**
  * Repository to read and write [Channel] data.
  */
 @SuppressWarnings("TooManyFunctions")
 internal class DatabaseChannelRepository(
+    private val scope: CoroutineScope,
     private val channelDao: ChannelDao,
     private val getUser: suspend (userId: String) -> User,
     private val getMessage: suspend (messageId: String) -> Message?,
     cacheSize: Int = 1000,
 ) : ChannelRepository {
 
     private val logger by taggedLogger("Chat:ChannelRepository")
+    private val channelCache = LruCache<String, Channel>(cacheSize)
 
-    /**
-     * Inserts a [Channel]
-     *
-     * @param channel [Channel]
-     */
-    override suspend fun upsertChannel(channel: Channel) {
-        val entity = channel.convertToEntity()
-        logger.v { "[upsertChannel] entity: ${entity.lastMessageInfo()}" }
-        channelDao.insert(entity)
+    override suspend fun insertChannel(channel: Channel) {
+        insertChannels(listOf(channel))
     }
 
     /**
      * Inserts many [Channel]s.
      *
      * @param channels collection of [Channel]
      */
-    override suspend fun upsertChannels(channels: Collection<Channel>) {
+    override suspend fun insertChannels(channels: Collection<Channel>) {
         if (channels.isEmpty()) return
-        val entities = channels.map { channel -> channel.convertToEntity() }
-        logger.v { "[upsertChannels] entities.size: ${entities.size}" }
-        channelDao.insertMany(entities)
+        val updatedChannels = channels
+            .map { channelCache[it.cid]?.let { cachedChannel -> it.combine(cachedChannel) } ?: it }
+        val channelToInsert = updatedChannels
+            .filter { channelCache[it.cid] != it }
+            .map { it.toEntity() }
+        cacheChannel(updatedChannels)
+        scope.launch {
+            logger.v {
+                "[insertChannels] inserting ${channelToInsert.size} entities on DB, " +
+                    "updated ${updatedChannels.size} on cache"
+            }
+            channelToInsert
+                .takeUnless { it.isEmpty() }
+                ?.let { channelDao.insertMany(it) }
+        }
```

Comment:
Hey @JcMinarro,

Wdyt about making this scope as a single-thread scope and put all write operations behind that scope?

It seems like we may have race-condition when we insert 1000 channels and delete 5 channels.
Let's save we had
```kotlin
val cidToDelete = "some_cid"
insertChannels(channelList)
// channelList contains cidToDelete
deleteChannel(cidToDelete)
```
In this case `deleteChannel ` will faster than `insertChannels`, so `insertChannels` will re-insert the deleted channel

Reply:
Let me check it

Reply:
Looks like single-thread scope won't handle it as we expect: https://stackoverflow.com/questions/61126844/coroutines-on-a-single-thread-and-guaranteed-sequentiality

Reply:
Maybe we could use `mutex` then to keep the order?

Reply:
Mutex I think is a safe bet, we can just something like this - since we only need to synchronise this one code block and we can synchronise it on the instance itself.

```
        synchronized(DatabaseChannelRepository@this) {
            ....
        }
````